Remove GetFileAttributesW for performance

shirosaki opened the pull request:
Thank you for timing result.
I tried rails startup bench with `to_long_name` = 1. Startup is faster if removing GetFileAttributesW at my benchmark.

According to Yura's cached-lp result, expand_path is called against load_path many times. And load_path is existing path. GetFileAttributesW is just overhead for existing path. So removing would be better when cached-lp is not enabled.

shirosaki committed:
Remove GetFileAttributesW for performance

Rails startup is faster when removing GetFileAttributesW.

luislavena commented:
Hello Hiroshi,

The following is the chart comparing original Fenix versus this modification and `to_long_name = 1`

http://cl.ly/0W2p21360x461P3z1W1x/hiroshi-2.png

Compared to previous one:

http://f.cl.ly/items/0g141e0V2g1D1i3C2p0O/ruby-hiroshi-fenix.png

Seems to me that the removal of `GetFileAttributesW` resulted in worse result for the synthetic benchmarks.

However, it scored similar when loading Rails in production mode (emptyapp)

Normal:

```
V:\emptyapp>timer ruby script\rails runner "0"
real    7.985
system  4.789
user    3.135
```

This pull:

```
V:\emptyapp>timer ruby script\rails runner "0"
real    6.690
system  3.884
user    2.792
```

Previous one:

```
V:\emptyapp>timer ruby script\rails runner "0"
real    6.890
system  4.040
user    2.730
```

However it is still twice slower than original Fenix:

```
V:\emptyapp>timer ruby script\rails runner "0"
real    3.828
system  1.092
user    2.698
```

`FindFirstFileW` seems to be the expensive call:

```
API                                                     Return                  Duration
GetFullPathNameW ( "bar\foo", 520, 0x0028e4dc, NULL )   19                      0.0000056
FindFirstFileW ( "V:\emptyapp\bar\foo", 0x0028e8ec )    INVALID_HANDLE_VALUE    0.0000789

API                                                         Return      Duration
GetFullPathNameW ( "script\rails", 520, 0x0028e4dc, NULL )  2           0.0000093
FindFirstFileW ( "V:\emptyapp\script\rails", 0x0028e8ec )   0x00700938  0.0002348
FindClose ( 0x00700938 )                                    TRUE        0.0000243
```

UPDATE: used spaces for the API call table.

shirosaki commented:
File.expand_path seems slow mainly cause of FindFirstFileW slowness.
I guess perhaps only skipping FindFirstFileW in MRI file.c would cause similar improvement as original Fenix.

https://github.com/ruby/ruby/blob/trunk/file.c#L3199

MRI calls FindFirstFileW against `dir` (second argument of expand_path) at recursive call, but this implementation doesn't. I think long name conversion of `dir` is not needed. This would be faster than MRI because of skipping FindFirstFileW of `dir`.

https://github.com/ruby/ruby/blob/trunk/file.c#L2965
https://github.com/ruby/ruby/blob/trunk/file.c#L2989

shirosaki committed:
Reduce object creation

Use fake String instead of object creation. It gives small performance
improvement.

shirosaki committed:
Add specs for UNC

Don't remove UNC root slashes.

shirosaki commented:
I've added two commits for other issues.
Please take a look. Thank you.
